lib/trino-orc/src/main/java/io/trino/orc/metadata/statistics/StringStatisticsBuilder.java

```diff
@@ -158,16 +176,82 @@ public ColumnStatistics buildColumnStatistics()
         return stringStatisticsBuilder.buildStringStatistics();
     }
 
-    private Slice dropStringMinMaxIfNecessary(Slice minOrMax)
+    private Slice computeStringMinMax(Slice minOrMax, boolean isMin)
     {
-        if (minOrMax == null || minOrMax.length() > stringStatisticsLimitInBytes) {
+        if (minOrMax == null || (!shouldCompactMinMax && minOrMax.length() > stringStatisticsLimitInBytes)) {
             return null;
         }
+        if (minOrMax.length() > stringStatisticsLimitInBytes) {
+            if (isMin) {
+                return StringCompactor.truncateMin(minOrMax, stringStatisticsLimitInBytes);
+            }
+            else {
+                return StringCompactor.truncateMax(minOrMax, stringStatisticsLimitInBytes);
+            }
+        }
 
         // Do not hold the entire slice where the actual stats could be small
         if (minOrMax.isCompact()) {
             return minOrMax;
         }
         return Slices.copyOf(minOrMax);
     }
+
+    static final class StringCompactor
+    {
+        private StringCompactor() {}
+
+        public static Slice truncateMin(Slice slice, int maxBytes)
+        {
+            checkArgument(slice.length() > maxBytes);
+            if (maxBytes == 0) {
+                return null;
+            }
+            int lastIndex = findLastCharacterInRange(slice, maxBytes);
+            return slice.slice(0, lastIndex);
```

Comment:
This can end up returning empty bytes array (empty slice), so why are we special-casing the maxBytes==0 case above

(eg consider an input being 4-byte utf8 sequence, and maxBytes=3; please add a test)

Reply:
So I had a check here for a situation when maxBytes is longer than provided byte array but you wrote that it is a dead code. It will obviously fell now for situations when slice is longer than maxBytes

Reply:
> So I had a check here for a situation when maxBytes is longer than provided byte array

yes, but that's not the situation i'm concerned about here